### PR TITLE
UI: Update Button types to allow for no children on iconOnly buttons

### PR DESCRIPTION
**Pull request by cdedreuille**
## What I did

I improved the types of the new Button component to allow for iconOnly buttons to not have children.

## How to test

1. Run this branch locally
2. Test the Button anywhere with `<Button iconOnly icon={<><>} />`
3. You should not see any typescript errors

## Checklist

- [x] Make sure your changes are tested (stories and/or unit, integration, or end-to-end tests)
- [x] Make sure to add/update documentation regarding your changes
- [x] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

#### Maintainers

- [x] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [x] Make sure this PR contains **one** of the labels below.

`["cleanup", "BREAKING CHANGE", "feature request", "bug", "build", "documentation", "maintenance", "dependencies", "other"]`

### 🦋 Canary release

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

### CODE REVIEW

File: code/ui/components/src/new/Button/Button.tsx [hunk 1]
```diff
@@ -4,18 +4,30 @@ import { styled } from '@storybook/theming';
...
   variant?: 'primary' | 'secondary' | 'tertiary';
-  icon?: ReactNode;
-  iconOnly?: boolean;
   onClick?: () => void;
   disabled?: boolean;
   active?: boolean;
 }
 
+type ButtonIconOnlyProps = {
+  iconOnly?: true;
```
Comment (kasperpeulen) [inline code/ui/components/src/new/Button/Button.tsx:17]: I think this should not be optional for the case.

Comment (kasperpeulen) [inline code/ui/components/src/new/Button/Button.tsx:17]: I think this should not be optional in this case

**Review by kasperpeulen — CHANGES_REQUESTED** [review]

File: code/ui/components/src/new/Button/Button.tsx (hunk 1 above)
Comment (cdedreuille) [reply code/ui/components/src/new/Button/Button.tsx:17]: Thanks. I'll make it mandatory.

**Review by kasperpeulen — APPROVED** [review]
LGTM!

**Pull request merged**